sympy/polys/numberfields.py

```diff
@@ -48,39 +50,50 @@ def _choose_factor(factors, x, v, dom=QQ, prec=200, bound=5):
     Return a factor having root ``v``
     It is assumed that one of the factors has root ``v``.
     """
+    from sympy.polys.polyutils import illegal
+
     if isinstance(factors[0], tuple):
         factors = [f[0] for f in factors]
     if len(factors) == 1:
         return factors[0]
 
-    points = {x:v}
+    prec1 = 10
+    points = {}
     symbols = dom.symbols if hasattr(dom, 'symbols') else []
-    t = QQ(1, 10)
-
-    for n in range(bound**len(symbols)):
-        prec1 = 10
-        n_temp = n
-        for s in symbols:
-            points[s] = n_temp % bound
-            n_temp = n_temp // bound
-
-        while True:
-            candidates = []
-            eps = t**(prec1 // 2)
-            for f in factors:
-                if abs(f.as_expr().evalf(prec1, points)) < eps:
-                    candidates.append(f)
-            if candidates:
-                factors = candidates
-            if len(factors) == 1:
-                return factors[0]
-            if prec1 > prec:
-                break
-            prec1 *= 2
+    while True:
+        # when dealing with non-Rational numbers we usually evaluate
+        # with `subs` argument but we only need a ballpark evaluation
+        xv = {x:v if not v.is_number else v.n(prec1)}
+        fe = [f.as_expr().xreplace(xv) for f in factors]
+
+        # assign integers [0, n) to symbols (if any)
+        for n in subsets(range(bound), k=len(symbols), repetition=True):
+            for s, i in zip(symbols, n):
+                points[s] = i
+
+            # evaluate the expression at these points
+            candidates = [(abs(f.subs(points).n(prec1)), i)
+                for i,f in enumerate(fe)]
+
+            # if we get invalid numbers (e.g. from division by zero)
+            # we try again
+            if any(i in illegal for i, _ in candidates):
+                continue
+
+            # find the smallest two -- if they differ significantly
+            # then we assume we have found the factor that becomes
+            # 0 when v is substituted into it
+            can = sorted(candidates)
+            (a, ix), (b, _) = can[:2]
+            if a < 10**6*b:  # XXX what to use?
```

Comment:
`a` is the smaller one:
```
> /home/jks/test/sympy/sympy/polys/numberfields.py(88)_choose_factor()
-> if a < 10**6*b:  # XXX what to use?
(Pdb) p a, b
(5.801108341e-11, 44.13886734)
```

Reply:
Oh yes. Surprising that it passes all the tests...

Reply:
It looks like that confirms my expectation that the minimum could always be chosen.

Reply:
If there are no symbols then I can see that the minimum could be chosen.

If we are substituting random integers for the symbols though then could it not be the case that multiple factors end up being exactly zero?

Reply:
That is possible though may not occur in our tests. So some condition like this is necessary.

Reply:
> If we are substituting random integers for the symbols though then could it not be the case that multiple factors end up being exactly zero?

There is a case when two zeros are computed in the current tests, but this was because of precision that was too low when I tested it. I forget if it happened at prec of 2 or 10. But the check that the smallest be uniquely small is what keeps the wrong root from being chosen.

I wish I could talk with a blackboard and chalk with you about this problem. Not being as versed in mathematical understanding, it is fascinating to me to see expressions that are supposed to be 0 *not be zero* for certain input. I could bring this up on the mailing list, if you think it better -- and I should probably add a wiki page about this since non-trivial zero expressions come up often enough. But for now, here is something I saw:

For one test the two factors are the Poly below:

```
[
(2.8, Poly(y, y, domain='QQ(x)')), 
(0.040, Poly(x**2*y**2 - 4*x**3 - 4, y, domain='QQ(x)'))]
```
The value for `x` being tested as being a root is
```
y = sqrt(x)/sqrt(1 + x**(-3)) - sqrt(x**3 + 1)/x + 1/(x**(5/2)*(1 + x**(-3))**(3/2)) + 1/(x**(11/2)*(1 + x**(-3))**(3/2))
```
Substituting into the two factors yields expressions that are a function of `x` and we need to decide which one is zero. *But I really don't understand what "zero" means in the context of `minpoly`.* Does it mean that it will be zero for Rational/Real input, only? Because if you substitute `x = I` you get
```
>>> z = y.subs(x, I); z.n()
0.910179721124455 + 2.19736822693562*I
>>> nsimplify(z)
sqrt(-2 + 2*sqrt(2)) + I*sqrt(2 + 2*sqrt(2))
>>> _.n()
0.910179721124455 + 2.19736822693562*I
>>> z.equals(nsimplify(z))
False  <--- that's wrong, but that is a different issue
```
So we can see that the expression we are testing to see if it is zero -- the factor with `y` replaced with a root -- is not identically 0 for arbitrary input. It *is* zero for real input, however, even input that gives imaginary roots:
```
>>> denoms(y, x)
{sqrt(1 + x**(-3)), x**(11/2), x**3, x**(5/2), x, (1 + x**(-3))**(3/2)}
>>> [i.subs(x,-S.Half) for i in _]
[sqrt(7)*I, -sqrt(2)*I/64, -1/8, sqrt(2)*I/8, -1/2, -7*sqrt(7)*I]
>>> y.subs(x, -S.Half)
0
```
So what does this test for satisfaction of a value being a root of a factor really mean?

Reply:
@asmeurer (thankful for comments you have made regarding such matters; perhaps you might care to comment here)

Reply:
The expression is:
```python
In [62]: yv = nsimplify(sqrt(x)/sqrt(1 + x**(-3)) - sqrt(x**3 + 1)/x + 1/(x**(5/2)*(1 + x**(-3))**(3/2)) + 1/(x**(11/2)*(1 + x**(-3))**(3/2)))

In [63]: yv
Out[63]: 
                   ________                                       
                  ╱  3                                            
      √x        ╲╱  x  + 1           1                   1        
───────────── - ─────────── + ──────────────── + ─────────────────
     ________        x                     3/2                 3/2
    ╱     1                    5/2 ⎛    1 ⎞       11/2 ⎛    1 ⎞   
   ╱  1 + ──                  x   ⋅⎜1 + ──⎟      x    ⋅⎜1 + ──⎟   
  ╱        3                       ⎜     3⎟            ⎜     3⎟   
╲╱        x                        ⎝    x ⎠            ⎝    x ⎠ 
```
SymPy computes this expression as zero for any real `x` with `x > -1`. For `x < -1` nonzero values are obtained:
```python
In [64]: yv.subs(x, -2)
Out[64]: √7⋅ⅈ
```
We can split the expression up into two terms taking the one with the sqrt in the numerator apart from the others:
```python
In [70]: y1 = yv.args[3]

In [71]: y2 = sum(yv.args[:3])

In [72]: y1
Out[72]: 
    ________ 
   ╱  3      
-╲╱  x  + 1  
─────────────
      x      

In [73]: y2
Out[73]: 
      √x               1                   1        
───────────── + ──────────────── + ─────────────────
     ________                3/2                 3/2
    ╱     1      5/2 ⎛    1 ⎞       11/2 ⎛    1 ⎞   
   ╱  1 + ──    x   ⋅⎜1 + ──⎟      x    ⋅⎜1 + ──⎟   
  ╱        3         ⎜     3⎟            ⎜     3⎟   
╲╱        x          ⎝    x ⎠            ⎝    x ⎠   
```
For `x > -1` the two parts `y1` and `y2` are nonzero but sum to zero. For `x < -1` both are imaginary but have the same sign so they combine rather than cancel. This depends on our choice of branch for the square roots I guess though so if we allowed `y1` to take the other branch then the terms would still cancel.

From purely algebraic considerations we can find a larger polynomial that this expression would need to be a root of e.g.:
```python
In [12]: from sympy.solvers.solvers import unrad

In [13]: factor(unrad(yv - y)[0])
Out[13]: 
                           14                   
   2        14 ⎛ 2        ⎞   ⎛   3    2  2    ⎞
x⋅y ⋅(x + 1)  ⋅⎝x  - x + 1⎠  ⋅⎝4⋅x  - x ⋅y  + 4⎠
```
This leaves 3 possible choices for `y`:
```python
In [14]: r1, r2, r3 = roots(_, y)

In [15]: r1.subs(x, I)
Out[15]: 
      ____________
2⋅ⅈ⋅╲╱ -ⅈ⋅(1 + ⅈ) 

In [16]: r1.subs(x, I).n()
Out[16]: 0.910179721124455 + 2.19736822693562⋅ⅈ

In [17]: r2.subs(x, I).n()
Out[17]: -0.910179721124455 - 2.19736822693562⋅ⅈ

In [18]: r3.subs(x, I).n()
Out[18]: 0
```
Exactly which one we have depends on our interpretation of the square root and other non-integer powers. The minpoly function substitutes only nonnegative integers for `x` and uses evalf/mpmath so all radicals take the principal root. That's why it chooses the zero root in this case. That's not the root it would find if it chose to substitute negative values for `x` though.

Reply:
I'm actually not really sure how well-defined this is. The correct choice of minimal polynomial for this case depends on the value of `x` but `x` here is just considered to be an arbitrary symbol.

Reply:
> The minpoly function substitutes only nonnegative integers for `x` and uses evalf/mpmath so all radicals take the principal root.

Is there some assurance that arguments to radicals are non-negative for non-negative, real values of `x`?

Reply:
> Is there some assurance that arguments to radicals are non-negative for non-negative, real values of `x`?

No, not really. I think that this depends essentially on the interpretation that we give for what `x` is supposed to mean in this expression and that isn't really specified anywhere.

Reply:
`y = sqrt(x)/sqrt(1 + x**(-3)) - sqrt(x**3 + 1)/x + 1/(x**(5/2)*(1 + x**(-3))**(3/2)) + 1/(x**(11/2)*(1 + x**(-3))**(3/2))` is an interesting example. It is the sum `y1 + y2` (https://github.com/sympy/sympy/pull/21370#discussion_r619889487) where `y1` is `-sqrt(x**3 + 1)/x` and `y2` can be simplified (manually) to `sqrt(x)*sqrt(1 + 1/x**3)`. The sum will be 0, if `sqrt(x**3 + 1)` is written in the form `x*sqrt(x)*sqrt(1 + 1/x**3)`. That cannot be done for all values of `x` as `sqrt(a*b)` is different from `sqrt(a)*sqrt(b)` in general.

If `y` were to denote an element of an algebraic function field (in `risch`, for example), then the field should be first constructed by giving suitable generators, `u = sqrt(x)` and `v = sqrt(1 + 1/x**3)`, say. Then `y` would be rewritten in terms of the generators. At that point, a choice would have to be made for `sqrt(x**3 + 1)`: either `x*u*v` or `-x*u*v`. Then the element would be unambiguous and its minimal polynomial well defined, either `y` or `x**2*y**2 - 4*x**3 - 4` (in monic form).

The algorithms in `numberfields` have been implement under the assumption that the elements lie in a well defined field. That is obviously true for numbers but not for expressions involving symbols. The minimal polynomial of a field element (over some base field like `QQ` or `QQ(x)`) is always irreducible. It might be possible to extend the definition of minimal polynomial to expressions like `y` above by taking it as the product of several irreducible factors but I'm not sure that it would be worth while.